lib/ansible/modules/system/macos_pref.py

```diff
@@ -0,0 +1,372 @@
+        self.return_value = self.current_value
+
+    def delete(self):
+        if self.current_value is None:
+            return
+
+        self.changed = True
+        self.return_value = None
+
+        if self.module.check_mode:
+            return
+
+        set_pref(self.key, None, self.domain)
+
+    def write(self):
+        if (isinstance(self.current_value, collections.MutableMapping)
+                and self.state == 'merge'):
+            new = copy.deepcopy(self.current_value)
+            deep_merge_dicts(new, self.value)
+        else:
+            new = self.value
+
+        if self.current_value == new:
+            return
+
+        self.changed = True
+        self.return_value = new
+
+        if self.module.check_mode:
+            return
+
+        if not set_pref(self.key, new, self.domain, kCFPreferencesCurrentUser, self._host_arg()):
+            self.success = False        
+    
+    def run(self):
+
+        self.read()
+
+        if self.state == "absent":
+            self.delete()
+
+        elif self.value is not None:
+            self.write()
+
+
+def deep_merge_dicts(base, incoming):
+    """
+    Performs an *in-place* deep-merge of key-values from :attr:`incoming`
+    into :attr:`base`. No attempt is made to preserve the original state of
+    the objects passed in as arguments.
+
+    :param dict base:  The target container for the merged values. This will
+        be modified *in-place*.
+    :type base:  Any :class:`dict`-like object
+
+    :param dict incoming:  The container from which incoming values will be
+        copied. Nested dicts in this will be modified.
+    :type incoming:  Any :class:`dict`-like object
+
+    :rtype:  None
+
+    """
+    for ki, vi in incoming.items():
+        if (
+            ki in base
+            and isinstance(vi, collections.MutableMapping)
+            and isinstance(base[ki], collections.MutableMapping)
+        ):
+            deep_merge_dicts(base[ki], vi)
+        else:
+            base[ki] = vi
+
+
+def get_pref(key, domain, username, hostname):
+    return pythonCollectionFromPropertyList(
+        CoreFoundation.CFPreferencesCopyValue(key, domain, username, hostname)
+    )
+
+
+def set_pref(key, value, domain, username, hostname):
+    CoreFoundation.CFPreferencesSetValue(key, value, domain, username, hostname)
+    return CoreFoundation.CFPreferencesAppSynchronize(domain)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            domain=dict(
+                default="NSGlobalDomain",
+                type='str',
+                required=False
+            ),
+            host=dict(
+                choices=[
+                    "anyHost",
+                    "currentHost"
+                ],
+                default="anyHost",
+                required=False
+            ),
+            key=dict(
+                type='str',
+                required=True
+            ),
+            value=dict(
```

Comment:
You should add an explicit code snippet below to verify that if state is replace/merge, the value must be provided. Have you tested what happens if state is set to replace/merge, but no value is provided?

Reply:
Per design if `value` is ommited, the module just returns (reads) the value of `key`. See the last example here https://github.com/ansible/ansible/pull/35683#issuecomment-364058748
In that example `state` is also ommited which means it defaults to `replace`.

You can find that logic here:
https://github.com/kinglouie/ansible/blob/0e23532dff2bad7ff4aa7a5429f58db8c5c72a72/lib/ansible/modules/system/macos_pref.py#L271-L279

I think it is simple enough to not confuse users.

Reply:
Hm... Yeah, you're right on that one. Although... I can actually now see why having `read` as a state could help reduce some confusion/improve consistency. For example, the following samples will read parameter in spite of having the explicit state that denotes changes to attributes:

```
- name: Show all files
  macos_pref:
    domain: com.apple.finder
    key: AppleShowAllFiles
    state: merge

- name: Show all files
  macos_pref:
    domain: com.apple.finder
    key: AppleShowAllFiles
    state: replace
```

A user that forgets to provide value might end-up thinking the invocations have actually changed the value. So, from usability perspective (at expense of being more expressive for the `read` operation), one could produce more explicit error information for the user.

Your call on this one, though :)

Reply:
I see your concerns but I think a user forgetting to provide a value when he wants to set something is rare enough it can be neglected in favor of fewer options to be set in 99% of the cases.